lib/parsers/ogg_parser.rb

```diff
@@ -0,0 +1,42 @@
+# https://xiph.org/vorbis/doc/Vorbis_I_spec.pdf
+# https://en.wikipedia.org/wiki/Ogg#Page_structure
+class FormatParser::OggParser
+  # The maximum size of an Ogg page is 65,307 bytes.
+  MAX_POSSIBLE_PAGE_SIZE = 65307
+
+  def call(io)
+    # The format consists of chunks of data each called an "Ogg page". Each page
+    # begins with the characters, "OggS", to identify the file as Ogg format.
+    capture_pattern = io.read(4)
+    return unless capture_pattern == 'OggS'
+
+    io.seek(28) # skip not important bytes
+
+    # Each header packet begins with the same header fields.
+    #   1) packet_type: 8 bit value (the identification header is type 1)
+    #   2) the characters v','o','r','b','i','s' as six octets
+    packet_type, vorbis, _vorbis_version, channels, sample_rate = io.read(16).unpack('Ca6LCL')
+    return unless packet_type == 1 && vorbis == 'vorbis'
+
+    # Read the last page of the audio in order to calculate the duration.
+    pos = io.size - MAX_POSSIBLE_PAGE_SIZE
+    pos = 0 if pos < 0
+    io.seek(pos)
+    page = io.read(MAX_POSSIBLE_PAGE_SIZE)
+    pos_of_the_last_page = page.rindex('OggS')
+    header = page[pos_of_the_last_page..pos_of_the_last_page + 13]
+
+    _capture_pattern, _version, _header_type, granule_position = header.unpack('a4CCQ')
```

Comment:
Is 13 the size of `a4CCQ`?

Reply:
And similar - `Q` is native-endian byte order. Absolute most formats specify endianness which can be hinted with the wider/newer `unpack` specifiers using `<` and `>`.

Reply:
> Is 13 the size of a4CCQ?

Is it wrong? I calculate it like this:
`Q` is 8 bytes
`C` is 1 byte
`a4` is 4 bytes

Reply:
Nope, nothing is "wrong" I was just double-checking since the `read` and the `unpack` are not very close to each other. This should be fine as long as you get enough bytes from your read()